### ISSUE TITLE
[BUG] OutOfMemoryError: Direct buffer memory when uploading large file

### ISSUE THREAD

**Issue by asm0dey** [Azure/azure-sdk-for-java#10823]
**Describe the bug**
When I'm trying to upload file much bigger than heap to blob storage size I'm getting OutOfMemoryError

***Exception or Stack Trace***
```
Operator called default onErrorDropped
java.lang.IllegalStateException: java.lang.OutOfMemoryError: Direct buffer memory
	at io.netty.channel.AbstractCoalescingBufferQueue.releaseAndCompleteAll(AbstractCoalescingBufferQueue.java:351)
	at io.netty.channel.AbstractCoalescingBufferQueue.releaseAndFailAll(AbstractCoalescingBufferQueue.java:207)
	at io.netty.handler.ssl.SslHandler.releaseAndFailAll(SslHandler.java:1814)
	at io.netty.handler.ssl.SslHandler.setHandshakeFailure(SslHandler.java:1793)
	at io.netty.handler.ssl.SslHandler.setHandshakeFailure(SslHandler.java:1757)
	at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:775)
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:748)
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:740)
	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:726)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.flush(CombinedChannelDuplexHandler.java:531)
	at io.netty.channel.ChannelOutboundHandlerAdapter.flush(ChannelOutboundHandlerAdapter.java:125)
	at io.netty.channel.CombinedChannelDuplexHandler.flush(CombinedChannelDuplexHandler.java:356)
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:748)
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:740)
	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:726)
	at reactor.netty.channel.MonoSendMany$SendManyInner.run(MonoSendMany.java:296)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.OutOfMemoryError: Direct buffer memory
	at java.base/java.nio.Bits.reserveMemory(Bits.java:175)
	at java.base/java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:118)
	at java.base/java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:317)
	at io.netty.buffer.PoolArena$DirectArena.allocateDirect(PoolArena.java:758)
	at io.netty.buffer.PoolArena$DirectArena.newChunk(PoolArena.java:734)
	at io.netty.buffer.PoolArena.allocateNormal(PoolArena.java:245)
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:227)
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:147)
	at io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:342)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:178)
	at io.netty.handler.ssl.SslHandler$SslEngineType$1.allocateWrapBuffer(SslHandler.java:215)
	at io.netty.handler.ssl.SslHandler.allocateOutNetBuf(SslHandler.java:2145)
	at io.netty.handler.ssl.SslHandler.wrap(SslHandler.java:821)
	at io.netty.handler.ssl.SslHandler.wrapAndFlush(SslHandler.java:792)
	at io.netty.handler.ssl.SslHandler.flush(SslHandler.java:773)
	... 17 more

```

**To Reproduce**

***Code Snippet***
```kotlin
azure.getBlobClient(blobName).uploadFromFile(absolutePath, overwrite)
```

**Expected behavior**
Do not fail with error

**Setup (please complete the following information):**
 - OS: Windows, Linux, Mac OS X
 - IDE : IntelliJ
 - Version of the Library used: 12.6.1

**Additional context**
I have bunch of another reports/questions because I have very special setup: I'm developing plugin for IDE, so my files are being loaded with another classloader (which is not even URLClassloader) and also it works inside IDE so potentially strange side-effects may occure.

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [X] Bug Description Added
- [X] Repro Steps Added
- [X] Setup information Added

**Comment by asm0dey** [comment]
Here you can find temporary link to heapdump: https://send.firefox.com/download/9d8d79c79476569b/#NV3srXLNSMtz3GhgPqoEDw

**Comment by asm0dey** [comment]
Here is minimal reproducable example:

```kotlin
fun main() {
    BlobServiceClientBuilder()
        .endpoint("https://pavelfinkelshteynjb.blob.core.windows.net")
        .credential(/* credentials here */)
        .buildClient()
        .getBlobContainerClient("testcontainer1")
        .getBlobClient("temp")
        .uploadFromFile(
            "/home/finkel/temp",
            ParallelTransferOptions(null, null, { println("Uploaded $it") }, null),
            null,
            null,
            null,
            null,
            null
        )
}
```

where `/home/finkel/temp` is 4-gigabyte file, created by fallocate.

**Comment by rickle-msft** [comment]
Hi, @asm0dey Thanks for reporting this. A few things. The file you shared is too large to be opened by any editors I have... Any way you could trim some of it down?  (This one may already be in the heap dump) What's the size of your heap? 

My hypothesis is there's too much concurrency going on in reading the file. I suspect that we're reading the file much faster than we are able to upload blocks and free that memory, which means too much of the file is being held in memory at once. If that's the case, which hopefully your heapdump can support once I can take a look, then we can potentially offer a way to limit the concurrency.

Another thing I'd be curious about is if you update your ParallelTransferOptions to shrink the block size if we might see that help a bit. With smaller blocks, uploads can complete more quickly and hopefully free up some memory to reduce pressure. This doesn't have to be a permanent solution, but it would again go towards validating this theory.

**Comment by asm0dey** [comment]
It's heap dump, not thread dump, so it can't be opened with editors, it should be opened with [Eclipse MAT](https://www.eclipse.org/mat/)/whatever tool you use for heap dump analyzing. My heap is smth like 1Gb.

I did try playing with Parallel transfer options, nothing helps.

**Comment by asm0dey** [comment]
But if this basic example doesn't work — does this functionality works at all?

**Comment by rickle-msft** [comment]
I'm quite confident uploadFromFile works. I just uploaded a 12GB from my machine and even limited the heap size to 1GB and didn't have any problems there. It could be something to do with your environment, but I'm not certain. 

@alzimmermsft It looks like a lot of the memory is being held onto by Netty. Any other thoughts on why it might not be releasing memory?

**Comment by asm0dey** [comment]
I can create minimal project that will reproduce the issue if you need.

**Comment by rickle-msft** [comment]
If you are able to do that, it would be very helpful for the investigation. Thank you!

**Comment by asm0dey** [comment]
[export.zip](https://github.com/Azure/azure-sdk-for-java/files/4595737/export.zip)
You can just change credentials and endpoint and run application. For me it fails. And for several of my colleagues too.

**Comment by asm0dey** [comment]
BTW, my java is:
```
openjdk version "13.0.2" 2020-01-14
OpenJDK Runtime Environment (build 13.0.2+8)
OpenJDK 64-Bit Server VM (build 13.0.2+8, mixed mode)
```

My OS is
```
Linux UNIT-1793 5.5.19-1-MANJARO #1 SMP PREEMPT Tue Apr 21 08:12:19 UTC 2020 x86_64 GNU/Linux
```

**Comment by rickle-msft** [comment]
@asm0dey Thank you for the repro. I think I was able to confirm that I think the file is being read too quickly. I actually was able to upload the 4gb file if I set the chunk size to 100kb. That's probably not something you want to keep permanently but might be a good stop gap. 

I think the actual fix is we need to give you guys a way to configure the level of parallelism on these operations. I was also able to upload successfully by limiting the number of concurrent operations happening in a patched version of the sdk, so if we allow customers to set that value, then you should also be able to experiment and find what works for your environment.

**Comment by asm0dey** [comment]
So does it reproduce for you in my example? If yes — than it's very basic scenario as I said. But I can't even imagine why did it work for you on 12gb file?

About level of parallelism — I believe there is such option, `ParallelTransferOptions#numBuffers`, it's lowest possible setting is `2`, but setting it doesn't help in my example.

**Comment by rickle-msft** [comment]
It does reproduce when I limit the size of the heap to 1GB. I was testing a 12GB file without limiting the size of the heap, so that is probably the difference. I think this hasn't shown up before because although we have done some memory profiling, we don't usually test under such memory constraints, and the combination of conditions under "normal" circumstances simply allows us to complete requests and free memory before this becomes an issue. 

We only use that option when doing a buffered upload, which is a different api. In that case we have to allocate new memory to copy the source into to make it safe for retries among other reasons. We don't do that copy when uploading from files because it isn't necessary, so that parameter gets ignored, as I believe is stated in the doc for that type. Although broadly, yes, in that case the number of buffers does affect the amount of parallelism because we can't issue more requests than the number of buffers we have. 

In this case, there is an overload on one of the methods we are using from the underlying async framework that we can take advantage of which actually is called `maxConcurrency` and explicitly limits the number of concurrent operations rather than implicitly because of resource constraints (i.e. numBuffers in the other case). I found that setting this to an arbitrary value of 5 worked for this example just as a proof of concept.

**Comment by asm0dey** [comment]
Oh, I get it. I believe that reactor should handle this under almost any memory limits — backpressure is all about that. Also, I believe that 1-2Gbs of heap is quite normal for end-users' computers. 

Do I get you right that I should use another API for uploading files? If yes — that using wrong API is indeed my fault. Which one should I use?

**Comment by rickle-msft** [comment]
Ideally it would handle it, but we have some custom code for reading from AsynchronousFileChannels, so it's possible backpressure isn't being handled properly at any point between reading the file all the way to getting it on the network. @anuchandy or @alzimmermsft can either of you comment on whether it sounds reasonable that we're reading from the source file faster than we're able to push data on the network and thus hitting memory problems? And if so, whether you agree that limiting the concurrency in our flatmap operation would be the right way to go?

No there should just be the one uploadFromFile method, so we'll fix this one and get it working for you. You shouldn't have to use different methods based on memory constraints :)

**Comment by zcox** [comment]
We are seeing a very similar error, but when downloading via `BlobClient.downloadToFile()`, using `azure-storage-blob` version `12.6.1`:

```
Caused by: io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 65536 byte(s) of direct memory (used: 52416116, max: 52428800)
```

We've observed that we have to set `-XX:MaxDirectMemorySize` to a size larger than the blob/file size, e.g. to download a 500MB file we must set something like `-XX:MaxDirectMemorySize=600M`.

Should I open a separate issue? Or include this as part of this issue?

**Comment by rickle-msft** [comment]
@zcox Thank you for chiming in here. At this point I think we can keep them in the same thread. Without falling into too much confirmation bias, I think the reverse of what I've described above is conceivable in the download case as well. If we are being too aggressive on parallelism, we'll issue too many concurrent requests and then be holding a bunch of that data in memory before it can be written to the file. Like I said, I don't want to assume these things have the same root cause, but it is possible.

**Comment by asm0dey** [comment]
Interesting detail is that be default you have as much direct memory allowed as big your allowed heap is. So uploading 12G-file will take max 24gb available to commit on your computer :)

**Comment by anuchandy** [comment]
@rickle-msft - piggybacking off part of the comment you left 

<i>
From: Rick

In this case, there is an overload on one of the methods we are using from the underlying async framework that we can take advantage of which actually is called "maxConcurrency" and explicitly limits the number of concurrent operations... "I found that setting this to an arbitrary value of 5 worked for this example just as a proof of concept".
</i>

the `maxConcurrency` parameter of flatMap operator (that is not configurable from Azure SDK) default to 256. One way to configure it is through the system property `reactor.bufferSize.small`. Here is the doc [reference](https://github.com/reactor/reactor-core/blob/a8022380f7d1eb9ef5bdcf0a9e0366d3dad534df/reactor-core/src/main/java/reactor/util/concurrent/Queues.java#L85) for this system property.

**Comment by rickle-msft** [comment]
@anuchandy Thanks for sharing that. I tried using xs yesterday, and I think I still hit oom errors. Would we be opposed to allowing customers configure this maxConcurrency directly on the api level?

**Comment by anuchandy** [comment]
@rickle-msft - agree with you on allowing the user to configure concurrency (maybe via `ParallelTransferOptions`) on the SDK API level. 

Between the XS_BUFFER_SIZE i.e. "reactor.bufferSize.x" controls the number of chunk-metadata values prefetched from the upstream file chunk-metadata [publisher](https://github.com/Azure/azure-sdk-for-java/blob/ef5bb972b5af148e9fe4f6f81efbb4e849fd4bf0/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java#L566). In my previous comment, I meant SMALL_BUFFER_SIZE i.e. the system property "reactor.bufferSize.small" which in a way controls the number of parallel call to [client.stageBlockWithResponse](https://github.com/Azure/azure-sdk-for-java/blob/ef5bb972b5af148e9fe4f6f81efbb4e849fd4bf0/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java#L539) that actually read part of File. Just want to make sure we're on the same page w.r.t the application of these two properties.

**Comment by anuchandy** [comment]
@rickle-msft @asm0dey - I tried uploading 4GB file with `-XX:MaxDirectMemorySize=512M` and as expected it failed with OOM. 

Then I tried setting both `concurrency` and `prefetch` to a lower value and 4 MB block-size like below (by keeping direct memory as it is i.e. 512MB):

```java
System.setProperty("reactor.bufferSize.small", "5")
System.setProperty("reactor.bufferSize.x", "5")

// The upload code here, blocksize set to 4MB
// ParallelTransferOptions(4 * 1024 * 1024, null, { println("Uploaded $it") }, null),
//    null,
//    null,
//    null,
//    null,
//    null)
```

with this upload seems working fine, its past 2 GB now. (It used OOM long before with default `concurrency` and `prefetch`).

So I believe Rick's hypothesis on too much concurrency is correct, and we should allow the users to configure concurrency to fine-tune it based on their environment. I'm also checking the direct memory allocation now.

@asm0dey can you give it a try in your environment? and if that works would you be able to use it as a short term solution to unblock your development?

**Comment by anuchandy** [comment]
as an another validation - along with direct memory I also set a limit on the heap memory. 
`-XX:MaxDirectMemorySize=256M -Xmx256M` where total memory (direct + heap) is 512 MB. The upload is functioning fine with these constraints as well, as long as we limit concurrency and prefetch.

```java
System.setProperty("reactor.bufferSize.small", "5") // limit concurrency
System.setProperty("reactor.bufferSize.x", "5")        // limit prfetch      

// The upload code here, blocksize set to 4MB
// ParallelTransferOptions(4 * 1024 * 1024, null, { println("Uploaded $it") }, null),
//    null,
//    null,
//    null,
//    null,
//    null)
```

**Comment by asm0dey** [comment]
I believe that default settings should be safe and conservative by default. Nothing should fail with default settings. Do you agree?

**Comment by roneiv** [comment]
@anuchandy 
There’s this open bug related to the reactor.bufferSize settings, so one should be careful what settings are applied and not:
[https://github.com/reactor/reactor-core/issues/1958](https://github.com/reactor/reactor-core/issues/1958)

**Comment by rickle-msft** [comment]
@asm0dey I agree that defaults should work in most cases out of the box. When we make the fix, we'll explore what some reasonable values might be. I think the conversation at this point is just validating that this will resolve the issue before we take the change. To that end, it'd be helpful if you could try setting the configuration that Anu shared to verify if it helps in your environment. 

@roneiv Thanks for sharing this! That's really helpful and we'll keep this in mind when we are implementing the fix.

**Comment by roneiv** [comment]
@rickle-msft Changing the bufferSize settings did not have an impact on the async dataLake client at least, we still experienced the same problem with memory leak and an OOM in the end.

**Comment by anuchandy** [comment]
@rickle-msft  probably the default the 256 for conucrrency is seems high. Maybe this shouldn't be aggressive but use a default value based on the number of connections available. Number of Connections is based on the number of cores.

At the same time, we should explore allowing the user to set the concurrency (i.e. the number of blocks to be uploaded in parallel) if override is desired.

**Comment by asm0dey** [comment]
@rickle-msft Setting these properties didn't help us too.

**Comment by anuchandy** [comment]
@asm0dey if possible could you check whether the flatMap is indeed picking these settings in the plugin environment? A breakpoint [here](https://github.com/reactor/reactor-core/blob/a8022380f7d1eb9ef5bdcf0a9e0366d3dad534df/reactor-core/src/main/java/reactor/core/publisher/FluxFlatMap.java#L72) should show picked value.

**Comment by asm0dey** [comment]
@anuchandy Nice catch! No, they don't :(

**Comment by anuchandy** [comment]
@asm0dey thanks for checking. Do you see 256 or 16 in plugin env? I noticed that the reactor won't allow setting concurrency lower than 16, so any value below we provide will result in 16.

**Comment by asm0dey** [comment]
Yes, that's what I see.

**Comment by rickle-msft** [comment]
@asm0dey Just to be clear, we're assuming you mean you're seeing 16. It's interesting that setting these properties doesn't fix anything as Anu and I are both seeing these system properties allow the upload to complete successfully. I'm using a 1000mb heap. 

Anu also found that we can try setting the system property `reactor.netty.pool.maxConnections` to work on limiting the connections as well. (The defaults are shown [here](https://github.com/reactor/reactor-netty/blob/master/src/main/java/reactor/netty/resources/ConnectionProvider.java#L56)). Can we try reducing the number of connections and see if that makes a difference?

Also just to double check something from earlier, you said you tried playing with your ParallelTransferOptions. I was able to get the transfer working with a block size of 100kb. Did you try a value in that range at all?

**Comment by asm0dey** [comment]
Nop, didn't try 100kb yet, will try.

And about sys props — I believe it can be all about classloaders and visibility. Or can't? You can easily check that I'm doing things right in simpish plugin from comment above: https://github.com/Azure/azure-sdk-for-java/issues/10823#issuecomment-625493785

**Comment by rickle-msft** [comment]
Another option here is if you're willing, you can download the branch where I limited the concurrency on this api specifically, install that, and try having your application pick up this dependency. The branch is [here](https://github.com/rickle-msft/azure-sdk-for-java/tree/oomError). It's set to version `12.7.0-beta.1.oom`. From the top of the repo, you should be able to run `mvn install -f pom.xml --fail-at-end -DskipTests -Dgpg.skip -pl "com.azure:azure-storage-blob" -am`. I was able to successfully run the sample project you gave using this dependency is is more or less what we'd change in the actual fix (except we'd make it configurable instead of hard coded).

**Comment by rickle-msft** [comment]
Thank you guys as well for being such active participants in resolving this bug. It is very helpful.

**Comment by rickle-msft** [comment]
@asm0dey FYI #11478

**Comment by asm0dey** [comment]
Thanks!

On 20/05/27 12:35PM, Rick Ley wrote:
> @asm0dey FYI #11478 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/Azure/azure-sdk-for-java/issues/10823#issuecomment-634895518

-- 
Regards,
Pasha

Big Data Tools @ JetBrains

**Comment by rickle-msft** [comment]
Hi, @asm0dey I have just released 12.7.0-beta.1 which should have a number of fixes for issues you've opened. Please give it a try and let us know so we can hopefully finalize before we GA those fixes next week. In the case of the OOM error you're hitting, you should be able to play with the maxConcurrency value on ParallelTransferOptions and see a difference. You might try setting it to something like 2 just to see if it works at all and then scale up from there. 

Please let us know how it works for you!

**Comment by asm0dey** [comment]
Yay!

Thank you so much! I will start testing it today :)

On 20/06/04 04:43PM, Rick Ley wrote:
> Hi, @asm0dey I have just released 12.7.0-beta.1 which should have a number of fixes for issues you've opened. Please give it a try and let us know so we can hopefully finalize before we GA those fixes next week. In the case of the OOM error you're hitting, you should be able to play with the maxConcurrency value on ParallelTransferOptions and see a difference. You might try setting it to something like 2 just to see if it works at all and then scale up from there. 
> 
> Please let us know how it works for you!
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/Azure/azure-sdk-for-java/issues/10823#issuecomment-639174093

-- 
Regards,
Pasha

Big Data Tools @ JetBrains

**Comment by asm0dey** [comment]
Thanks! 2 works for us and I think we'll stick with it because you can never predict in which environment your customer runs IDE :)

**Comment by rickle-msft** [comment]
Makes sense to me! Thank you for giving it a test! I am going to close this issue since I believe it has been resolved. The fix is in master and will go out with our GA release this week or next.

**Comment by d-schmidt** [comment]
According to the reactor [Queues](https://github.com/reactor/reactor-core/blob/a8022380f7d1eb9ef5bdcf0a9e0366d3dad534df/reactor-core/src/main/java/reactor/util/concurrent/Queues.java#L79) class, this is the minimum:

        System.setProperty("reactor.bufferSize.small", "16"); // default 256
        System.setProperty("reactor.bufferSize.x", "8"); // default 32
